titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/KubernetesNodeDataResolver.java

```diff
@@ -0,0 +1,115 @@
+    private final Predicate<V1Node> nodeFilter;
+
+    private final Function<String, Matcher> relocationRequiredTaintsMatcher;
+    private final Function<String, Matcher> relocationRequiredImmediatelyTaintsMatcher;
+
+    public KubernetesNodeDataResolver(RelocationConfiguration configuration,
+                                      KubeApiFacade kubeApiFacade,
+                                      Predicate<V1Node> nodeFilter) {
+        this.nodeInformer = kubeApiFacade.getNodeInformer();
+        this.relocationRequiredTaintsMatcher = RegExpExt.dynamicMatcher(
+                configuration::getNodeRelocationRequiredTaints,
+                "nodeRelocationRequiredTaints",
+                Pattern.DOTALL,
+                logger);
+        this.relocationRequiredImmediatelyTaintsMatcher = RegExpExt.dynamicMatcher(
+                configuration::getNodeRelocationRequiredImmediatelyTaints,
+                "nodeRelocationRequiredImmediatelyTaints",
+                Pattern.DOTALL,
+                logger);
+        this.nodeFilter = nodeFilter;
+    }
+
+    @Override
+    public Map<String, Node> resolve() {
+        List<V1Node> k8Nodes = nodeInformer.getIndexer().list().stream().filter(nodeFilter).collect(Collectors.toList());
+        Map<String, Node> result = new HashMap<>();
+        k8Nodes.forEach(k8Node -> toReconcilerNode(k8Node).ifPresent(node -> result.put(node.getId(), node)));
+        return result;
+    }
+
+    private Optional<Node> toReconcilerNode(V1Node k8Node) {
+        if (k8Node.getMetadata() == null
+                || k8Node.getMetadata().getName() == null
+                || k8Node.getMetadata().getLabels() == null
+                || k8Node.getSpec() == null
+                || k8Node.getSpec().getTaints() == null) {
+            return Optional.empty();
+        }
+
+        Map<String, String> k8Labels = k8Node.getMetadata().getLabels();
+
+        String serverGroupId = k8Labels.get(NODE_LABEL_MACHINE_GROUP);
+        if (serverGroupId == null) {
+            return Optional.empty();
+        }
+
+        Node node = Node.newBuilder()
+                .withId(k8Node.getMetadata().getName())
+                .withServerGroupId(serverGroupId)
+                .withRelocationRequired(anyNoExecuteMatch(k8Node, relocationRequiredTaintsMatcher))
+                .withRelocationRequiredImmediately(anyNoExecuteMatch(k8Node, relocationRequiredImmediatelyTaintsMatcher))
+                .build();
+        return Optional.of(node);
+    }
+
+    private boolean anyNoExecuteMatch(V1Node k8Node, Function<String, Matcher> taintsMatcher) {
+        return k8Node.getSpec().getTaints().stream().anyMatch(taint ->
+                TAINT_EFFECT_NO_EXECUTE.equals(taint.getEffect()) && taintsMatcher.apply(taint.getKey()).matches()
+        );
+    }
+
+    /**
+     * Kubernetes informer does not provide staleness details, just information about the first sync.
+     */
+    @Override
+    public long getStalenessMs() {
```

Comment:
thoughts on changing the contract here to return `Optional<Long>`, and be more explicit when sync has not occurred?

Reply:
We use this approach in other places/APIs, so to be consistent we should change it everywhere.